node/network/availability-distribution/src/lib.rs

```diff
@@ -579,18 +575,16 @@ where
 /// Obtain the first key which has a signing key.
 /// Returns the index within the validator set as `ValidatorIndex`, if there exists one,
 /// otherwise, `None` is returned.
-fn obtain_our_validator_index(
+async fn obtain_our_validator_index(
 	validators: &[ValidatorId],
-	keystore: KeyStorePtr,
+	keystore: SyncCryptoStorePtr,
 ) -> Option<ValidatorIndex> {
-	let keystore = keystore.read();
-	validators.iter().enumerate().find_map(|(idx, validator)| {
-		if keystore.has_keys(&[(validator.to_raw_vec(), PARACHAIN_KEY_TYPE_ID)]) {
-			Some(idx as ValidatorIndex)
-		} else {
-			None
+	for (idx, validator) in validators.iter().enumerate() {
+		if CryptoStore::has_keys(&*keystore, &[(validator.to_raw_vec(), PARACHAIN_KEY_TYPE_ID)]).await {
```

Comment:
This will be absolutely slow in the moment we have remote signing. Sounds like we need a function that returns all the indices of the keys that it has.

Reply:
So we only make one remote call.

Reply:
This needs a change in the way that `has_keys` works where it should return the index of the items found from the array sent as an argument. This could be implemented as a follow-up PR to both substrate and polkadot. What do you think?

Reply:
https://github.com/paritytech/substrate/issues/7285

Reply:
Yeah, the idea was to fix this in a follow up pr.